### PR TITLE
Refactor Numeric literal conversion and add tests.

### PR DESCRIPTION
**Pull request by associahedron**
The logic around LF conversion of numeric literals was a bit sketchy, so in this PR I tried to refactor it (based on Martin's review of the last PR) and to handle polymorphic literals more carefully. I added a positive test for numeric literals, and some negative tests.

### CODE REVIEW

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 1]
```diff
@@ -237,36 +237,34 @@ convertRationalDecimal env num denom
...
-                (num, denom)
+convertRationalNumericMono env scale num denom
+    | scale < 0 || scale > fromIntegral numericMaxScale =
+        unsupported
+            ("Tried to construct value of type Numeric " ++ show scale ++ ", but scale is out of bounds. Scale must be between 0 through 37, not " ++ show scale ++ ".")
+            scale
+
+    | abs (rational * 10 ^ scale) >= 10 ^ numericMaxPrecision =
+        unsupported
+            ("Rational is out of bounds: " ++ show double ++ ". The range of values representable by the Numeric " ++ show scale ++ " type is  -10^" ++ show maxPower ++ " + 1  through  10^" ++ show maxPower ++ " - 1.")
```
Comment (hurryabit) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: Shouldn't this be something like
```suggestion
            ("Rational is out of bounds: " ++ show double ++ ". The range of values representable by the Numeric " ++ show scale ++ " type is  -10^" ++ show maxPower ++ " + 10^-" ++ show scale ++ "  through  10^" ++ show maxPower ++ " - 1.")
```

Comment (associahedron) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: ah, you're right

Comment (associahedron) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: gosh, what a cumbersome number to describe ... perhaps I can say "Numeric scale can only represent numbers greater than -10^maxPower and smaller than 10^maxPower"

**Comment by associahedron** [comment]
I added some negative tests for cases that weren't already covered by the existing Rational/Decimal tests.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 1 above)
Comment (hurryabit) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: I would use language like "between ... (exlusive)" or "strictly between". But I don't really care.

**Comment by remyhaemmerle-da** [comment]
This PR advances the state of #2289.

**Pull request merged**